### PR TITLE
Cuda request gpu device expr

### PR DESCRIPTION
**Pull request by tetron**
* Revise CUDARequirement so to support expressions to choose how many GPUs we want.
* adjust cudaComputeCapability to take an array of explicit versions to use

### CODE REVIEW

File: cwltool/extensions-v1.1.yml [hunk 1]
```diff
@@ -93,13 +93,24 @@ $graph:
...
+          capabilities that explicitly appear in the array.
+    cudaDeviceCountMin:
+      type: ['null', int, cwl:Expression]
       default: 1
       doc: Minimum number of GPU devices to request, default 1.
-    deviceCountMax:
-      type: int?
+    cudaDeviceCountMax:
+      type: ['null', int, cwl:Expression]
       doc: Maximum number of GPU devices to request.  If not specified, same as `deviceCountMin`.
```
Comment (mr-c) [inline cwltool/extensions-v1.1.yml:116]:
```suggestion
      doc: Maximum number of GPU devices to request.  If not specified, same as `cudaDeviceCountMin`.
```

File: cwltool/extensions.yml [hunk 2]
```diff
@@ -203,13 +203,24 @@ $graph:
...
+          capabilities that explicitly appear in the array.
+    cudaDeviceCountMin:
+      type: ['null', int, cwl:Expression]
       default: 1
       doc: Minimum number of GPU devices to request, default 1.
-    deviceCountMax:
-      type: int?
+    cudaDeviceCountMax:
+      type: ['null', int, cwl:Expression]
       doc: Maximum number of GPU devices to request.  If not specified, same as `deviceCountMin`.
```
Comment (mr-c) [inline cwltool/extensions.yml:226]:
```suggestion
      doc: Maximum number of GPU devices to request.  If not specified, same as `cudaDeviceCountMin`.
```

File: cwltool/job.py [hunk 3]
```diff
@@ -180,7 +181,9 @@ def _setup(self, runtimeContext: RuntimeContext) -> None:
             "http://commonwl.org/cwltool#CUDARequirement"
         )
         if cuda_req:
-            count = cuda_check(cuda_req)
+            count = cuda_check(
```
Comment (mr-c) [inline cwltool/job.py:184]: Needs testing coverage

File: cwltool/process.py [hunk 4]
```diff
@@ -995,20 +996,34 @@ def evalResources(
             "outdirMin": 1024,
             "outdirMax": 1024,
         }
-        for a in ("cores", "ram", "tmpdir", "outdir"):
+
+        cudaReq, _ = self.get_requirement("http://commonwl.org/cwltool#CUDARequirement")
+        if cudaReq:
+            request["cudaDeviceCountMin"] = 1
+            request["cudaDeviceCountMax"] = 1
```
Comment (mr-c) [inline cwltool/process.py:1003]: Lacks test coverage

File: cwltool/process.py [hunk 5]
```diff
@@ -1022,13 +1037,18 @@ def evalResources(
...
-        return {
+
+        defaultReq = {
             "cores": request_evaluated["coresMin"],
             "ram": math.ceil(request_evaluated["ramMin"]),
             "tmpdirSize": math.ceil(request_evaluated["tmpdirMin"]),
             "outdirSize": math.ceil(request_evaluated["outdirMin"]),
         }
+        if cudaReq:
+            defaultReq["cudaDeviceCount"] = request_evaluated["cudaDeviceCountMin"]
```
Comment (mr-c) [inline cwltool/process.py:1050]: Lacks test coverage

File: tests/wf/nvidia-smi-container.cwl [hunk 6]
```diff
@@ -5,7 +5,7 @@ $namespaces:
 requirements:
   cwltool:CUDARequirement:
     cudaVersionMin: "1.0"
-    cudaComputeCapabilityMin: "1.0"
+    cudaComputeCapability: "1.0"
```
Comment (mr-c) [inline tests/wf/nvidia-smi-container.cwl:8]: Probably needs a variant with the array version as well

**Review by mr-c — CHANGES_REQUESTED** [review]

File: tests/test_cuda.py [hunk 7]
```diff
@@ -1,10 +1,24 @@
...
 from cwltool.main import main
+from cwltool.pathmapper import MapperEnt, PathMapper
+from cwltool.process import use_custom_schema, use_standard_schema
+from cwltool.stdfsaccess import StdFsAccess
+from cwltool.update import INTERNAL_VERSION, ORIGINAL_CWLVERSION
+from cwltool.utils import CWLObjectType
 
 from .util import get_data, needs_docker, needs_singularity_3_or_newer
 
+from unittest.mock import MagicMock
```
Comment (mr-c) [inline tests/test_cuda.py:20]: Maybe use https://docs.pytest.org/en/7.0.x/how-to/monkeypatch.html ?

Comment (tetron) [reply tests/test_cuda.py:20]: I added `types-mock` to mypy-requirements and that makes mypy happy

**Review by mr-c — APPROVED** [review]

**Pull request merged**